api/v1beta1/machinedrainrules_types.go

```diff
@@ -0,0 +1,213 @@
+/*
+Copyright 2024 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package v1beta1
+
+import (
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+const (
+	// PodDrainLabel is the label that can be set on Pods in workload clusters to ensure a Pod is not drained.
+	// The only valid value is "skip".
+	// This label takes precedence over MachineDrainRules defined in the management cluster.
+	PodDrainLabel = "cluster.x-k8s.io/drain"
+)
+
+// MachineDrainRuleDrainBehavior defines the drain behavior. Can be either "Drain" or "Skip".
+// +kubebuilder:validation:Enum=Drain;Skip
+type MachineDrainRuleDrainBehavior string
+
+const (
+	// MachineDrainRuleDrainBehaviorDrain means a Pod should be drained.
+	MachineDrainRuleDrainBehaviorDrain MachineDrainRuleDrainBehavior = "Drain"
+
+	// MachineDrainRuleDrainBehaviorSkip means the drain for a Pod should be skipped.
+	MachineDrainRuleDrainBehaviorSkip MachineDrainRuleDrainBehavior = "Skip"
+)
+
+// MachineDrainRuleSpec defines the spec of a MachineDrainRule.
+type MachineDrainRuleSpec struct {
+	// drain configures if and how Pods are drained.
+	// +required
+	Drain MachineDrainRuleDrainConfig `json:"drain"`
+
+	// machines defines to which Machines this MachineDrainRule should be applied.
+	//
+	// If machines is empty, the MachineDrainRule applies to all Machines in the Namespace.
+	// If machines contains multiple selectors, the results are ORed.
+	// Within a single Machine selector the results of selector and clusterSelector are ANDed.
+	// Machines will be selected from all Clusters in the Namespace unless otherwise
+	// restricted with the clusterSelector.
+	// An empty Machine selector matches all Machines in the Namespace.
+	// This field is required.
+	//
+	// Example: Selects control plane Machines in all Clusters and
+	//          Machines with label "os" == "linux" in Clusters with label
+	//          "stage" == "production".
+	//
+	//  - selector:
+	//      matchExpressions:
+	//      - key: cluster.x-k8s.io/control-plane
+	//        operator: Exists
+	//  - selector:
+	//      matchLabels:
+	//        os: linux
+	//    clusterSelector:
+	//      matchExpressions:
+	//      - key: stage
+	//        operator: In
+	//        values:
+	//        - production
+	//
+	// +required
+	// +kubebuilder:validation:MinItems=1
+	// +kubebuilder:validation:MaxItems=32
+	Machines []MachineDrainRuleMachineSelector `json:"machines"`
```

Comment:
Should objects in this list be unique?

Reply:
> Do you think this is something that we should block?

Generally I would advise to prevent people from setting duplicate items in case there's a bug, that somehow always adds a new, identical item, without realising that there's an existing identical item in the list (yes, I have seen that, it was a fun one to work out)

> I assume I can block this with +kubebuilder:validation:UniqueItems, right? Then why not.

I think `UniqueItems` is actually broken isn't it? I always end up using

```
// +kubebuilder:validation:XValidation:rule="self.all(x, self.exists_one(y, x == y))",message="entries in machines must be unique"
```